Fixed #5676: Sum cost by quantity

Computroniks opened the pull request:
# Description

Currently the total cost displayed on the components, accessories, consumables and licences does not take into account the quantity of the item. An new function has been created based upon the sumFormatter function that takes into account the quantity of an item when calculating the total cost. It takes the purchase cost and multiplies it by the quantity column before adding it to the total.

Fixes #5676 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

1. Create a new component, accessory, consumable or licence with a quantity of 1 and any purchase cost
2. Take note of the total at the bottom ![chrome_w9XviNtB8B](https://user-images.githubusercontent.com/67638596/128265262-de421adb-2207-46e1-a354-b87c5c1ac105.png)
3. Increase the quantity of the item and note the total
![chrome_BXEEyVtANl](https://user-images.githubusercontent.com/67638596/128265360-428a9626-9028-4312-abce-440d416e1291.png)



**Test Configuration**:
* PHP version: 7.3.29-1~deb10u1
* MySQL version: 10.3.29-MariaDB-0+deb10u1
* Webserver version: Apache/2.4.38
* OS version: Raspbian GNU/Linux 10


# Checklist:

- [x] I have read the Contributing documentation available here: https://snipe-it.readme.io/docs/contributing-overview
- [x] I have formatted this PR according to the project guidelines: https://snipe-it.readme.io/docs/contributing-overview#pull-request-guidelines
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

Computroniks committed:
Added function to calculate cost based on quantity

sumFormatterQuantity takes the same input as sumFormatter but instead
of calculating the specified columns total it calculates the total
purchase cost of an item based upon its quantity. Also updated affected
pressenters to use this formatter.

Signed-off-by: Computroniks <mnickson@sidingsmedia.com>

Computroniks committed:
Licence cost calculation

Licences use diffrent key to track quantity. sumFormatterQuantity has
been modified to detect which key to use.

Signed-off-by: Computroniks <mnickson@sidingsmedia.com>

NikoGrano commented:
@snipe Sorry for pinging you directly here, but I've been following this and #5676 enough long without seeing any progress. 

Could it be possible to review this and merge if it's possible. Or give some comment about this at least.

snipe commented:
@uberbrady has been doing a lot of work on the JS/backend for currency recently, so I'd need him to look at this before I'd be willing to merge.

Computroniks commented:
Just looked at the commit for fixing the merge conflicts. Seems there is an issue with some of the lines getting duplicated. I will fix that now.

Computroniks committed:
Merge branch 'develop' into sum_cost_by_quantity

Computroniks commented:
Fixed the merge conflicts.

NikoGrano commented:
@uberbrady Would you mind to take look into this issue. Would be nice to get your input as there is active contributor to fix the problems - if any.

Computroniks commented:
Anything I can do to help get this merged?

uberbrady commented:
Yes - can you switch over to the European currency format? e.g. 1.234,56 - and see if it still works - and include screenshots?

We're slammed with the upcoming release of v6 and if you gave me something like that I might just take it and merge it, and deal with the consequences.

Computroniks committed:
Fixed sumFormatterQuantity if using 1.234,56 fomat

Previously sumFormatterQuantity used the parseFloat to convert the
string purchase_cost to a floating point number. parseFloat does not
return the correct value when using the comma format. To fix this
sumFormatterQuantity now used the cleanFloat function to convert
purchase_cost to a float.

Signed-off-by: Matthew Nickson <mnickson@sidingsmedia.com>

Computroniks commented:
Ok, there was a slight issue when using the comma format. I've just switched out the parse used to convert purchase_cost to a float from parseFloat to cleanFloat and it now works. Here are the screenshots. Not all of the components can be seen in the screenshots so the numbers will appear to not add up. This is just because not everything can be seen. The important bit is that the two numbers match as we already know that the function works when we use the decimal format.

Decimal format:

![Screenshot from 2021-10-30 22-49-38](https://user-images.githubusercontent.com/67638596/139559346-83a11f9c-e7dc-4337-a877-e9b13b7198c0.png)

and the comma format:

![Screenshot from 2021-10-30 22-50-18](https://user-images.githubusercontent.com/67638596/139559359-12b13e13-e30b-4166-8f6e-d39617b4a0b8.png)

They match! :)

Computroniks committed:
Merge branch 'develop' into sum_cost_by_quantity

Computroniks commented:
Fixed the merge conflicts

NikoGrano commented:
Ping @uberbrady

uberbrady commented on resources/views/partials/bootstrap-table.blade.php:
```diff
@@ -653,6 +653,31 @@ function sumFormatter(data) {
+            // Check that we are actually trying to sum cost from a table
+            // that has a quantity column. We must perform this check to
+            // support licences which use seats instead of qty
+            if('qty' in data[0]) {
```
This is very clever, I love it!

uberbrady reviewed (approved):
Sorry for the delays - this looks really good to me, so we'll be happy to take it.

I particularly love the way of figuring out the `multiplier` - very nice :)
